File: cacheMiss.go
package main

import (
	"fmt"
	"time"
)

const size = 20000

var final [size][size]byte

func main() {
	var t = time.Now()
	var arr [size][size]byte

	t = time.Now()
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			arr[j][i] = byte(i + j)
		}
	}
	final = arr
	fmt.Println(time.Now().Sub(t))

	t = time.Now()
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			arr[i][j] = byte(i + j)
		}
	}
	final = arr
	fmt.Println(time.Now().Sub(t))
}

File: cache-miss.clj
(def size 2000)

(def arr
  (vec (map vec
            (for [i (range 0 size)]
              (for [j (range 0 size)]
                (rand-int 5))))))

(defn profile []
  (println
   (time
    (let [ret
          (doall
           (for [i (range 0 size)]
             (doall
              (for [j (range 0 size)]
                (+ (nth (nth arr j) i) i j)))))])))
  (println
   (time
    (let [ret
          (doall
           (for [i (range 0 size)]
             (doall
              (for [j (range 0 size)]
                (+ (nth (nth arr i) j) i j)))))]))))
